# Assignment-1: Computational Genomics
## Task 1.1
A dataset "dataset.h5ad" has been provided, which contains a single-cell gene expression matrix. The following are the tasks to be performed on this dataset. 
1. Quality control, normalization, feature selection and PCA.
2. Following this, data is to be clustered using the Leiden Clustering Algorithm. 
3. Differential expression analysis is to be performed upon each cluster to identify the marker genes.  
4. Based on the marker genes identified, the clusters obtained are to be annotated. 

Note - the dataset can be accessed from the following link: [Large-Files](https://drive.google.com/drive/folders/1e78HYxcL1Z_czrz-b37Mc8F3HIk6b6th?usp=sharing). Let us begin this task by importing the required packages. 

### Loading the Dataset

In [3]:
import scanpy as sc 
import pandas as pd
import numpy as np

In [17]:
matrix = sc.read('C:/Personal/Academics/IITK Resources/Sem-7/Computational-Genomics/Large-Files/dataset.h5ad')
matrix

AnnData object with n_obs × n_vars = 10727 × 12303

What does this output mean? 
- n_obs refers to the number of cells. Hence, here we have 10,727 cells. 
- n_vars refers to the number of genes. This matrix contains details about 12,303 genes. 

Inspecting the AnnData object. 

In [18]:
matrix.var_names

Index(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
       ...
       '12293', '12294', '12295', '12296', '12297', '12298', '12299', '12300',
       '12301', '12302'],
      dtype='object', length=12303)

In [19]:
matrix.obs_names

Index(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
       ...
       '10717', '10718', '10719', '10720', '10721', '10722', '10723', '10724',
       '10725', '10726'],
      dtype='object', length=10727)

Adding the gene IDs provided.

In [20]:
gene_IDs = pd.read_csv('gene_names.csv')
gene_IDs

0
0      LINC00115
1         FAM41C
2         SAMD11
3          NOC2L
4         KLHL17
...          ...
12298       YBEY
12299       PCNT
12300      DIP2A
12301      S100B
12302      PRMT2

[12303 rows x 1 columns]

In [21]:
matrix.var_names = gene_IDs['0'].values
matrix.var_names

Index(['LINC00115', 'FAM41C', 'SAMD11', 'NOC2L', 'KLHL17', 'PLEKHN1', 'HES4',
       'ISG15', 'AGRN', 'TTLL10',
       ...
       'COL6A2', 'FTCD', 'LSS', 'MCM3AP-AS1', 'MCM3AP', 'YBEY', 'PCNT',
       'DIP2A', 'S100B', 'PRMT2'],
      dtype='object', length=12303)